`pulumi convert` generates invalid types for some resources

Given the following valid YAML program:

```yaml
name: yaml-test
runtime: yaml

configuration:
  domain:
    type: String
    default: example.com
  subdomain:
    type: String
    default: www

variables:
  domainName: ${subdomain}.${domain}
  zone:
    Fn::Invoke:
      Function: aws:route53:getZone
      Arguments:
        name: ${domain}

resources:

  certificate:
    type: aws:acm:Certificate
    properties:
      domainName: ${domainName}
      validationMethod: DNS

  certificateValidation:
    type: aws:route53:Record
    properties:
      name: ${certificate.domainValidationOptions[0].resourceRecordName}
      type: ${certificate.domainValidationOptions[0].resourceRecordType}
      zoneId: ${zone.zoneId}
      ttl: 60
      records:
        - ${certificate.domainValidationOptions[0].resourceRecordValue}
```

... `pulumi convert` fails because of a codegen error:

```
$ pulumi convert --out my-go-program --language go

main.go:56:81: expected operand, found 'return'
main.go:61:1: missing ',' in composite literal
main.go:63:1: expected operand, found '}'
main.go:65:1: expected ';', found 'if'
```

Internal thread for some additional context: https://pulumi.slack.com/archives/C027PR8QZPH/p1662664787503099

--- a/CHANGELOG_PENDING.md
+++ b/CHANGELOG_PENDING.md
@@ -2,3 +2,5 @@
 
 ### Bug Fixes
 
+- Fix invalid resource type on `pulumi convert` to Go
+  [#10670](https://github.com/pulumi/pulumi/pull/10670)

--- a/pkg/codegen/go/gen_program.go
+++ b/pkg/codegen/go/gen_program.go
@@ -924,7 +924,6 @@ func (g *generator) useLookupInvokeForm(token string) bool {
 // is aliased, returning that alias if available.
 func (g *generator) getModOrAlias(pkg, mod, originalMod string) string {
 	info, ok := g.getGoPackageInfo(pkg)
-	mod = strings.Split(mod, "/")[0]
 	if !ok {
 		return mod
 	}
@@ -940,6 +939,7 @@ func (g *generator) getModOrAlias(pkg, mod, originalMod string) string {
 			return path.Base(info.ImportBasePath)
 		}
 	}
+	mod = strings.Split(mod, "/")[0]
 	return mod
 }
 

--- a/pkg/codegen/go/gen_program.go
+++ b/pkg/codegen/go/gen_program.go
@@ -924,6 +924,7 @@ func (g *generator) useLookupInvokeForm(token string) bool {
 // is aliased, returning that alias if available.
 func (g *generator) getModOrAlias(pkg, mod, originalMod string) string {
 	info, ok := g.getGoPackageInfo(pkg)
+	mod = strings.Split(mod, "/")[0]
 	if !ok {
 		return mod
 	}

--- a/pkg/codegen/go/gen_program_expressions.go
+++ b/pkg/codegen/go/gen_program_expressions.go
@@ -184,8 +184,9 @@ func (g *generator) GenFunctionCallExpression(w io.Writer, expr *model.FunctionC
 					"Unsafe enum conversions from type %s not implemented yet: %s => %s",
 					from.Type(), from, to))
 			}
-			enumTag := fmt.Sprintf("%s.%s",
-				tokenToModule(to.Token), tokenToName(to.Token))
+			pkg, mod, typ, _ := pcl.DecomposeToken(to.Token, to.SyntaxNode().Range())
+			mod = g.getModOrAlias(pkg, mod, mod)
+			enumTag := fmt.Sprintf("%s.%s", mod, typ)
 			if isOutput {
 				g.Fgenf(w,
 					"%.v.ApplyT(func(x *%[3]s) %[2]s { return %[2]s(*x) }).(%[2]sOutput)",

